### PR TITLE
feat: Add Sharding Dimension Support to Parallel API

### LINKED ISSUE #345: [FEATURE] Add Sharding Position Support for Torch Parallel
# Feature Request

In the current Parallel API, users cannot specify the target dimension along which a tensor should be sharded. In PyTorch, this is done using Shard(dim) to indicate the sharding dimension. To bring similar functionality to Mithril, we propose extending the sharding specification format.

Currently, sharding is defined as (1, 4) meaning 4-way split along axis 1. We propose supporting an extended form like ((0, 1), (1, 4)), where:

- The first element of each tuple denotes the tensor dimension (e.g. 0 or 1),

- The second element denotes the sharding size along that dimension.

This would allow more explicit control over how tensors are partitioned.

## Motivation
Currently, the Parallel backend abstracts away the details of how tensors are sharded across devices. However, for advanced use cases—especially in distributed training and custom model parallelism—users may want to control which dimension is sharded. Lack of this control limits expressiveness and alignment with Torch-style workflows.

### PR DESCRIPTION
**Pull request by aturker-synnada**
## Description

Closes #345.

## What is Changed

-  Extended the Parallel API to accept explicit sharding dimension specification using a tuple of (dim, shards) pairs (e.g., ((0, 2), (1, 4)))
-  Preserved backward compatibility with the original (dim, shards) format

## Checklist:

- [x] Tests that cover the code added.
- [x] Corresponding changes documented.
- [x] All tests passed.
- [x] The code linted and styled (pre-commit run --all-files has passed).

### CODE REVIEW

File: mithril/backends/with_autograd/torch_backend/parallel.py [hunk 1]
```diff
@@ -199,34 +209,30 @@ def parallelize(  # type: ignore[override]
         self,
         tensor: torch.Tensor,
         base_mesh: DeviceMesh,
-        device_mesh: tuple[int, ...] | None = None,
+        device_mesh: tuple[int, ...] | tuple[tuple[int, int], ...] | None = None,
     ) -> STensor:
+        # TODO: The name `device_mesh` is suck, it should be renamed
```
Comment (kberat-synnada) [inline mithril/backends/with_autograd/torch_backend/parallel.py:214]: "# TODO: Rename `device_mesh` argument"

**Review by norhan-synnada — APPROVED** [review]
